src/arbitrary.rs

```diff
@@ -388,6 +392,63 @@ impl<T: Arbitrary> Arbitrary for VecDeque<T> {
     }
 }
 
+impl Arbitrary for PathBuf {
+    fn arbitrary<G: Gen>(g: &mut G) -> PathBuf {
+        // use some real directories as guesses, so we may end up with
+        // actual working directories in case that is relevant.
+        let here = std::env::current_dir().unwrap_or(PathBuf::from("/test/directory"));
+        let temp = std::env::temp_dir();
+        let home = std::env::home_dir().unwrap_or(PathBuf::from("/home/user"));
+        let mut p = g.choose(&[here,
+                               temp,
+                               home,
+                               PathBuf::from("."),
+                               PathBuf::from(".."),
+                               PathBuf::from("../../.."),
+                               PathBuf::new()]).unwrap().clone();
+        for sub in Vec::<OsString>::arbitrary(g).iter() {
+            p.push(sub);
+        }
+        p
+    }
+
+    fn shrink(&self) -> Box<Iterator<Item=PathBuf>> {
+        let mut shrunk = Vec::new();
+
+        let mut popped = self.clone();
+        if popped.pop() {
+            shrunk.push(popped);
+        }
+
+        // Iterating over a Path performs a small amount of normalization.
+        let normalized = self.iter().collect::<PathBuf>();
+        if normalized.as_os_str() != self.as_os_str() {
+            shrunk.push(normalized);
+        }
+
+        // Add the canonicalized variant only if canonicalizing the path actually does something,
+        // making it (hopefully) smaller. Also, ignore canonicalization if canonicalization errors.
+        if let Ok(canonicalized) = self.canonicalize() {
```

Comment:
If `popped.pop()` returns `false`, what is the behavior of `canonicalize`?

Reply:
Frankly, I have no idea. I think it could vary. If popping takes the path down to `/` or some other root, `canonicalize` should return the same path; if popping removes the path entirely (no root) who knows what canonicalize would do - probably error. So, `canonicalize` (I think) has multiple possible behaviors in the case of `!popped.pop()`.

Reply:
Hmm OK, I think my primary concern is whether this bottoms out or not. e.g., It `canonicalize` keeps returning different paths, then it's possible that the shrinker never terminates.

Reply:
(But that seems like a quite surprising behavior, so I think we're fine as is.)

Reply:
Hmm. I think it would be impossible to get stuck in a loop between `pop` and `canonicalize`. In order for `canonicalize` to normalize back to a different path, `pop` would need to pop back to a path which, when canonicalized, contains the original canonicalized path as a parent. `canonicalize` will result in wildly different paths when symbolic links are involved, and *should* remove things like `..` [citation needed]. So, I don't see a situation without a crazy looping hard link that will cause `canonicalize` and `pop` to cycle.